sdks/java/io/contextualtextio/src/main/java/org/apache/beam/sdk/io/contextualtextio/ContextualTextIO.java

```
@@ -423,90 +422,91 @@ public void processElement(
-      PCollectionView<Map<KV<String, Long>, Long>> numRecordsBeforeEachRange;
+      PCollectionView<Map<String, Map<Long, Long>>> numRecordsBeforeEachRange;
 
       public AssignRecordNums(
-          PCollectionView<Map<KV<String, Long>, Long>> numRecordsBeforeEachRange) {
+          PCollectionView<Map<String, Map<Long, Long>>> numRecordsBeforeEachRange) {
         this.numRecordsBeforeEachRange = numRecordsBeforeEachRange;
       }
 
       @ProcessElement
       public void processElement(ProcessContext p) {
-        Long range = p.element().getKey().getValue();
         String file = p.element().getKey().getKey();
+        Long offset = p.element().getKey().getValue();
         Row record = p.element().getValue();
-        Long numRecordsLessThanThisRange =
-            p.sideInput(numRecordsBeforeEachRange).get(KV.of(file, range));
+
+        Map<Long, Long> numRecordsBeforeEachOffsetInFile =
+            p.sideInput(numRecordsBeforeEachRange).get(file);
+        Long numRecordsLessThanThisOffset = numRecordsBeforeEachOffsetInFile.get(offset);
+
         Row newLine =
             Row.fromRow(record)
                 .withFieldValue(
                     RecordWithMetadata.RECORD_NUM,
                     record.getInt64(RecordWithMetadata.RECORD_NUM_IN_OFFSET)
-                        + numRecordsLessThanThisRange)
+                        + numRecordsLessThanThisOffset)
                 .build();
         p.output(newLine);
       }
+
+      private Long getNumRecordsBeforeOffset(
+          Long offset, Iterator<KV<Long, Long>> numRecordsBeforeEachOffsetInFile) {
+        while (numRecordsBeforeEachOffsetInFile.hasNext()) {
+          KV<Long, Long> entry = numRecordsBeforeEachOffsetInFile.next();
+          if (entry.getKey().equals(offset)) {
+            return entry.getValue();
+          }
+        }
+        LOG.error("Unable to compute contextual metadata. Please report a bug in ContextualTextIO");
+        return null;
```

Should we throw an exception here to fail the job ? Might be difficult for the user to debug the null in their metadata.